File: LeetCode_Python/LinkedList/DeleteNode/Leetcode_1669_merge_in_between_linked_lists/Leetcode_1669_merge_in_between_linked_lists.py
"""
@brief: Leetcode_1669_合并两个链表
@link: https://leetcode.cn/problems/merge-in-between-linked-lists/description/
@author: liangj.zhang
@date: 2/11/2025

@Difficulty: Medium

@Label: Delete Node

@Retrospect(worthy 1 - 5): 2

@thoughts:
    + 【思路 1】：在 list1 当中找到下标 a 的前一个节点，找到下标 b 的后一个节点；
        + 条件: 1 <= a <= b < list1.length - 1
            ==> 不需要考虑边界，[a, b] 范围严格在 (0, list1.length) 之间；
    
    + 解题过程：
        1. 遍历 list1，找到位置 a - 1 的节点 pre；
        2. 再遍历到位置 b + 1 的节点 next；
        3. 将 pre->next 指向 list2；
        4. 找到 list2 的末尾节点，将其 next 指向 next；
        5. 返回 list1。

    + 分析：
        + 时间复杂度：O(b + n)，其中 b 为索引值，n 为 list2 的长度
        + 空间复杂度：O(1)
    + rank:
        + 时间效率：206 ms, 击败 84.94%
        + 空间效率：22.29 MB, 击败 44.32%
"""

import os
import sys

print(os.path.join(os.path.dirname(__file__)))
print(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../.."))) #f:\Code\zljgit\LeetCode
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../..")))

from Include.PyLeetcode.LinkedList.LinkedList import ListNode

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeInBetween(self, list1: ListNode, a: int, b: int, list2: ListNode) -> ListNode:
        cur = list1
        pre = next = None
        i = 0

        while i + 1 < a:
            i += 1
            cur = cur.next
        pre = cur
        while i + 1 <= b + 1:
            i = i + 1
            cur = cur.next
        next = cur

        pre.next = list2
        while pre.next: 
            pre = pre.next
        pre.next = next

        return list1

File: LeetCode/LinkedList/DeleteNode/Leetcode_1669_merge_in_between_linked_lists/Leetcode_1669_merge_in_between_linked_lists.cpp
/**
 * @brief: Leetcode_1669_合并两个链表
 * @link: https://leetcode.cn/problems/merge-in-between-linked-lists/description/
 * @author: liangj.zhang
 * @date: 2/11/2025
 * 
 * @Difficulty: Medium
 * 
 * @Label: Delete Node
 * 
 * @Retrospect(worthy 1 - 5): 2
 * 
 * @thoughts:
 *  + 【思路 1】：在 list1 当中找到 a 下表索引的前一个节点，找到 b 下标索引的后一个节点；
 *      + condition: 1 <= a <= b < list1.length - 1     ==> 我们不必考虑边界，[a,b] 的范围是介于 (0, list1.length) 范围之内的；
 *      
 *      + 分析：
 *          + 时间复杂度：O(b + n), b -> 下表索引值, n -> list2 的长度
 *          + 空间复杂度：O(1)
 *      + rank:
 *          + 时间效率：199 ms, 击败 93.51%
 *          + 空间效率：97.14 MB, 击败 72.99%
 */

#include "../../../../Include/Leetcode/LinkedList/LinkedList.h"
#include <iostream>
#include <vector>
using std::vector;
using std::cout;
using std::endl;
using Leetcode::LinkedList::ListNode;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeInBetween(ListNode* list1, int a, int b, ListNode* list2) {
        
        ListNode* cur = list1;
        ListNode *pre(NULL), *next(NULL);
        int i = 0;

        while (++i < a) cur = cur->next;
        pre = cur;
        while (++i <= b + 1) cur = cur->next;
        next = cur->next;

        pre->next = list2;
        while (pre->next) pre = pre->next;
        pre->next = next;

        return list1;
    }
};